### ISSUE TITLE
Custom Mapping for List of Tuples not working?

### ISSUE THREAD

**Issue by joceron** [typelevel/doobie#2113]
I'm trying to insert values into a column of ClickHouse that has type `Array(Tuple(String, String))`, and on my Scala code it's a `List[(String, String)]`. I tried defining my custom mapping, but it seems that is being mapped to a `ClickHouseTupleValue` with a single value that contains the whole thing?.

This is how I created my table:
```sql
CREATE TABLE tmp (my_column Array(Tuple(String, String))) PRIMARY KEY my_column;
```

And this is the unit test where I'm trying to insert a value into that table:
```scala
import cats.data.NonEmptyList
import cats.effect._
import cats.effect.unsafe.implicits.global
import cats.implicits._
import doobie.Put
import doobie.hikari.HikariTransactor
import doobie.implicits._
import doobie.util.ExecutionContexts
import doobie.util.update.Update
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class DoobieSpec extends AnyWordSpec with Matchers {

  case class MyClass(my_column: List[(String, String)])

  implicit val listTuplePut: Put[List[(String, String)]] =
    Put.Advanced.array[(String, String)](NonEmptyList.one("Array"), "Tuple(String, String)").contramap(_.toArray)

  "DoobieSpec" should {
    "work" in {
      val query = "insert into tmp (my_column) values (?)"

      val items: Vector[MyClass] = Vector(MyClass(List(("first", "second"))))

      val transactorResource = for {
        transactionPool <- ExecutionContexts.fixedThreadPool[IO](16)
        transactor      <- HikariTransactor.newHikariTransactor[IO](
                               "com.clickhouse.jdbc.ClickHouseDriver"
                             , ???
                             , ???
                             , ???
                             , transactionPool
                             , None
                           )
      } yield transactor

      transactorResource
        .use(transactor => Update[MyClass](query).updateMany(items).transact(transactor))
        .unsafeRunSync()
    }
  }

}
```

That code throws me this exception:
```
Index: 1, Size: 1
java.lang.IndexOutOfBoundsException: 
	at java.base/java.util.Collections$SingletonList.get(Collections.java:4959)
	at com.clickhouse.data.format.ClickHouseRowBinaryProcessor$TupleSerializer.serialize(ClickHouseRowBinaryProcessor.java:243)
```

I tried to debug it, and I found out that the error is thrown on ClickHouse's Java Client, in `ClickHouseRowBinaryProcessor`. When trying to call `TupleSerializer.serialize`:
```java
public void serialize(ClickHouseValue value, ClickHouseOutputStream output) throws IOException {
    List<Object> tupleValues = value.asTuple();
    int i = 0;

    for(int len = this.serializers.length; i < len; ++i) {
        this.serializers[i].serialize(this.values[i].update(tupleValues.get(i)), output);
    }

}
```

That `value.asTuple`, I would expect it to return a List with the two items of my tuple, but it returns a single value with both items combined:
![Screenshot 2024-10-09 at 14 41 15](https://github.com/user-attachments/assets/8acdd4e1-8336-456a-9329-3eec1ab64ec5)

I know the exception is thrown at Java Client's, but I suppose the trigger was the way I defined that `listTuplePut: Put[List[(String, String)]]` on my unit test? Am I doing something wrong when defining that `Put` and I misunderstood the documentation, or is something on Doobie's side?

**Comment by jatcwang** [comment]
Hi @joceron. JDBC's array support is a bit driver-specific so I'm not quite sure off the top of my head. 

My recommendation is to check Clickhouse JDBC's documentation on how to set array values (particularly, how array of tuple can be set), get a pure JDBC version working and then compare it with the code in Put#array and see if the Put#array code matches up https://github.com/tpolecat/doobie/blob/main/modules/core/src/main/scala/doobie/util/put.scala#L181

If you need additional custom JDBC code, you will need to write your own Put#array implementation

**Comment by joceron** [comment]
Thanks @jatcwang ! I think I was able to reproduce the same situation using only code from the Java JDBC driver, and I opened an issue on their repository. I'll wait for an answer there, and then come back here with the resolution, so if possible I would like to keep the issue opened here until then

**Comment by joceron** [comment]
So, thanks to the help I got in the repository of Java JDBC driver, I was able to make it work. 

For future reference, in case someone else also have this issue: the trick is to not insert the data as an array of tuples, but as an array of arrays. And the inner array needs to have size 2. Then, if the column on ClickHouse has been defined as `Array(Tuple(_, _))`, it will be inserted with the right type. I don't know if I have to thank ClickHouse or the Java JDBC Driver for this.

So, the working implementation of the `Put` I defined in my example is this:
```scala
implicit val listTuplePut: Put[List[(String, String)]] = Put
  .Advanced
  .array[Array[String]](NonEmptyList.one("Array"), "Tuple(String, String)")
  .contramap(_.map { case (first, second) => List(first, second).toArray }.toArray)
```

And thanks again @jatcwang for the input. I'll close this issue as completed
